refactor - Calendar input

ibelar opened the pull request:
Use flatPickr js.

#### BC-Break
Small BC-Break if using Fomantic-UI specific options. You may have to set them using flatPickr
see: https://flatpickr.js.org/options/

Fix #1467 

<img width="370" alt="Screen Shot 2020-10-03 at 9 38 42 AM" src="https://user-images.githubusercontent.com/2204478/94993061-4ce2e380-055c-11eb-856f-4b1359f879a1.png">

<img width="356" alt="Screen Shot 2020-10-03 at 9 38 52 AM" src="https://user-images.githubusercontent.com/2204478/94993068-579d7880-055c-11eb-93fa-5d9817ee0ea0.png">

ibelar committed:
refactor - Calendar input

Use flatPickr js.

PhilippGrashoff commented:
@ibelar Whats your reason for not using FUI calendar input any more? In the last year, it was extended a lot and has a lot of functionality now.

Update: I just read #1467, this seems sensible reasoning. However, not having 2 different JS libs for the same inputs (FUI and SUI-Vue which are both used for input type="text" depending on Form Field for example) would be better I think. But we discussed this already and I understood that this would mean loooots of work.

ibelar committed:
add localization

ibelar committed:
allow time format changes and late options changed

ibelar committed:
Merge branch 'develop' into refactor/calendar

ibelar committed:
refactor init

mvorisek commented on src/Form/Control/Calendar.php:
```diff
@@ -4,78 +4,102 @@
+            case 'time':
+                $this->options['enableTime'] = true;
+                $this->options['noCalendar'] = true;
+                $this->options['time_24hr'] = $this->use24HrTime;
```
no property for 24h - parse php date format like:

```
private function is24hrTimeFormat(): bool
{
    return !preg_match('~[gh]~', $this->options['dateFormat'] ?? $this->app->ui_persistence->{$this->type . '_format'});
}
```

ibelar replied:
this is not for date parsing but rather an option for flatpickr on how to display/set time for user.

mvorisek replied:
if we set somewhere in php a format "option on how to display/set time for user", then parse if 24hr format from it, that is all

ibelar replied:
I think they are independent (setting vs output). But I will check and change accordingly. I will also need to add tests for this.

ibelar replied:
option is now set according to php format and test is added.

ibelar committed:
Merge branch 'develop' into refactor/calendar

ibelar committed:
switch to flatpickr vue component

- multiine now use flatpickr
- queybuilder now use flatpickr

ibelar commented on src/Form/Control/ScopeBuilder.php:
```diff
@@ -261,8 +262,51 @@ class ScopeBuilder extends Control
+                'enableTime' => true,
+                'noCalendar' => true,
+                'time_24hr' => true,
+            ],
```
@georgehristov - Can you take a look at how datetime and time rules are defined. Time and Datetime can now by supported via vue-flatpickr .
Question: 
Do we need datetime at all?
Flatpickr vue component needs to have proper set of properties define to switch from date, time or datetime picker. I add them via $rule['props'] property. Is that ok? Do we need to switch format based on ui persistence? if yes, then where do we set the property. If not, should we just set them in javascript not in php?

ibelar replied:
I remove props and now use rule['options'] see new comments.

ibelar committed:
refactor UI persistence

ibelar committed:
Merge branch 'develop' into refactor/calendar

ibelar committed:
fix getApp

ibelar committed:
move Date props to format array

ibelar committed:
fix

ibelar committed:
move type class into array container

mvorisek commented on src/Form/Control/Calendar.php:
```diff
@@ -146,4 +145,27 @@ public function jsGetFlatPickr(): JsExpression
+
+    private function use24hrTimeFormat(): bool
+    {
+        return !(bool) preg_match('~[gGh]~', $this->options['altFormat'] ?? $this->options['dateFormat']);
```
no need for bool cast

ibelar replied:
done!

mvorisek commented on src/Persistence/Type/Basic.php:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Base type class.
+ */
+class Basic
```
This dupliates a lot of data Type, not sure, if this is step into the right direction.

ibelar replied:
The idea was to:
- clearly separate each field type value casting to it's own class, i.e. clearer of what each type value casting does; 
- easily add a new field type without having to change Persistence/UI class;
- override each type class easily to adapt for custom needs;

mvorisek replied:
add/override is the only plus, but it strong argument to do it so.

I added some design comments directly

/cc @georgehristov

mvorisek commented on src/Persistence/Type/Castable.php:
```diff
@@ -0,0 +1,20 @@
+    /**
+     * Cast value when loaded from POST request.
+     */
+    public static function castLoadValue(Field $field, $value);
```
not static!

ibelar replied:
done!

mvorisek commented on src/Persistence/Ui.php:
```diff
@@ -19,35 +23,26 @@
+        'object' => Serial::class,
+    ];
+
+    public function getTypeClass(string $type): ?string
```
`getType(string $type): Castable`

ibelar replied:
done!

mvorisek commented on src/Persistence/Ui.php:
```diff
@@ -19,35 +23,26 @@
+    }
 
-    public $calendar_options = [];
+    public function setTypeClass(string $type, string $class)
```
registerTypeClass or addTypeClass

see how ORM/DBAL type registry works (yes, they are for php <->, but usage should be about the same)

ibelar replied:
done! I used registerTypeClass

ibelar committed:
refactor type

ibelar committed:
update comment / cs fixer

ibelar committed:
fixes

ibelar committed:
add sample

ibelar committed:
cs fixe

ibelar committed:
improve example

mvorisek commented on demos/form-control/calendar.php:
```diff
@@ -35,32 +35,32 @@
     'caption' => 'Allow input (F d, Y)',
-    'options' => ['allowInput' => true]])
-     ->set(date(Date::getFormat('date')));
+    'options' => ['allowInput' => true], ])
```
end `]` on new line, if there is NL after `[`

ibelar replied:
fix

ibelar committed:
more sample

ibelar committed:
fix label

ibelar committed:
cs fix

ibelar committed:
add behat test

ibelar committed:
cs fix

ibelar committed:
Merge branch 'develop' into refactor/calendar

ibelar committed:
fixes

ibelar committed:
update demo

ibelar committed:
fixes

ibelar committed:
fix getTheirModel to createTheirModel

ibelar committed:
Merge branch 'develop' into refactor/calendar

ibelar committed:
add flatpickr after merging develop

ibelar committed:
update cdn cloudflare

mvorisek commented on demos/form-control/calendar.php:
```diff
@@ -0,0 +1,80 @@
+$form = Form::addTo($layout, [], ['r1c1']);
+
+Date::setFormat('date', 'Y-m-d');
+$form->addControl('date_y_m_d', [Form\Control\Calendar::class, 'type' => 'date', 'caption' => 'Date (Y-m-d)'])
```
UI format types must be implemented in non-static way, probably once per App which we can then easily access with like `->frontFormatter` and conversion must be definitely handled in Control itself - in php/end user/programmer code, we must NEVER work on formatter data - these are only for user

mvorisek replied:
example here: for date type, type will allow to set `\DateTimeInterface` only, never formatted string from php

```
->set(new \DateTime());
```

ibelar replied:
Revert to previous ui_persistence for now.

mvorisek committed:
Increase memory theshold for burn testing

mvorisek committed:
Merge branch 'develop' into refactor/calendar

ibelar committed:
revert to previous ui_persistence

ibelar committed:
remove old comments

ibelar committed:
fix demo date input

ibelar committed:
refactor query builder props

ibelar commented on src/Form/Control/ScopeBuilder.php:
```diff
@@ -333,17 +335,43 @@ protected function buildQuery(Model $model)
         return $this;
     }
 
+    protected function getTypeOptions(string $type)
```
@georgehristov - I refactored ScopeBuilder using this rule['options'] instead of settings props to date, datetime or time type. This way, if we decide to use other Vue components in QueryBuilder, then we could simply pass component props using rule.options. 

I have also set dateFormat accoding to ui_persistence, which means that value sent to the server is format accordingly. Let me know if you need to change, i.e. if value sent always need to be in the same format.

ibelar replied:
@georgehristov - Just refactor (again) to use component type instead of field type. and change $rule['options'] to $rule['componentProps'] I think it make more sense this way.

ibelar replied:
Now use altFormat for date display and dedicated format for setting string value.

ibelar committed:
cs fix

mvorisek commented on src/Form/Control/Calendar.php:
```diff
@@ -145,26 +145,26 @@ public function getJsInstance(): JsExpression
     }
 
-    private function translateFormat(string $format): string
+    public static function translateFormat(string $format): string
```
is there any reason for static except "it does not require `$this` and maybe is expected to not change"?

ibelar replied:
SopeBuilder also uses flatpickr and I uses it to properly set options to it.

```
$options['dateComponent']['time_24hr'] = Calendar::use24hrTimeFormat($format);
$options['dateComponent']['noCalendar'] = ($type === 'time');
$options['dateComponent']['enableSeconds'] = Calendar::useSeconds($format);
```

mvorisek replied:
instantiate dummy Calendar control class in ScopeBuilder or UI Persistence if needed

ibelar replied:
and why not use static function? Options are not going to change. if they do, they will need to be change everywhere...

mvorisek replied:
Static anything is not optimal from OOP design in general as DI is not easy then. I understand, that here you expect it will not be required to change, but what if user want to use different Calendar lib? I am very happy, that Atk does not use static, so I want to keep that clean code.

ibelar replied:
make sense. I reverted it to public function instead.

ibelar committed:
use Calendar translate format

ibelar committed:
base props on component instead of type

ibelar committed:
remove static function

ibelar committed:
revert options

ibelar committed:
allow flatpickr option

ibelar committed:
change function name and add comments

ibelar committed:
cs fix

ibelar committed:
Merge branch 'develop' into refactor/calendar

ibelar committed:
enhancement

- query builder date, datetime or time always output same format but still display according to ui persistence

ibelar committed:
cs fix

ibelar committed:
support for condition value

ibelar committed:
use flatpickr altFormat

ibelar committed:
incl build

ibelar committed:
Merge branch 'develop' into refactor/calendar

ibelar committed:
set html for behat test

ibelar committed:
behat test for scopebuilder

ibelar committed:
behat fix

ibelar committed:
fix typo

ibelar committed:
more typo

ibelar committed:
add boolean rule

ibelar committed:
behat check for word match

ibelar committed:
rename demo to Scope Builder

mvorisek commented on tests-behat/bootstrap/Context.php:
```diff
@@ -428,6 +429,86 @@ public function iClickIconUsingCss($arg1)
+    }
+
+    /**
+     * @Then /^hasRef rule "([^"]*)" operator is "([^"]*)" and value is "([^"]*)"$/
```
Then ScopeBuilder rule...

and name method like scopeBuilderHasRef...

and not need for comment `//// ScopeBuilder rule /////////` above

ibelar committed:
clean up

ibelar committed:
remove comments

mvorisek commented on demos/_unit-test/scope-builder.php:
```diff
@@ -15,20 +17,33 @@
 $project = new Condition('project_name', Condition::OPERATOR_REGEXP, '[a-zA-Z]');
 // Display: Client Country Iso equals Brazil => ToWord: Client Country Iso is equal to 'Brazil'
-$brazil = new Condition('client_country_iso', '=', 'Brazil');
+$brazil = new Condition('client_country_iso', Condition::OPERATOR_EQUALS, 'Brazil');
```
well, I think `=` was invented for humans and well standardized, @georgehristov wdyt? Is there even any sense making this constant public?

ibelar replied:
What is the status here? should we use condition const or symbol?

georgehristov replied:
We should use symbols I would say

ibelar replied:
thanks, I will change it back to symbol.

ibelar committed:
update comments

ibelar committed:
improve function var

ibelar committed:
add check for input value in scopebuilder test

ibelar committed:
revert preg_replace

ibelar committed:
dump word

ibelar committed:
dump exprected with no dump

ibelar committed:
test

ibelar committed:
test no input

ibelar committed:
add container not found exeption

ibelar committed:
increase headless window size

ibelar committed:
update preg_replace pattern

ibelar committed:
improve input error message

ibelar committed:
revert to operator symbol

ibelar committed:
refactor mulitline flatpickr option

ibelar committed:
Merge branch 'develop' into refactor/calendar

mvorisek commented on tests-behat/bootstrap/Context.php:
```diff
@@ -500,7 +500,7 @@ public function iCheckIfWordMatch()
         $responseWord = $wordResponseContainer->getText();
 
-        if (preg_replace('/[^A-Za-z0-9\-]/', '', $expectedWord) !== preg_replace('/[^A-Za-z0-9\-]/', '', $responseWord)) {
+        if (preg_replace('/\s*/m', '', $expectedWord) !== preg_replace('/\s*/m', '', $responseWord)) {
```
```suggestion
        if (preg_replace('~\s*~', '', $expectedWord) !== preg_replace('~\s*~', '', $responseWord)) {
```

I highly prefer `~` as regex delimiter. `m` flag completely not needed as `\s` match NL even then.

ibelar replied:
ok - will try.

mvorisek commented on behat.yml.dist:
```diff
@@ -4,7 +4,7 @@ default:
         features: '%paths.base%/tests-behat'
       contexts:
-        - atk4\ui\behat\ContextDump
+        - atk4\ui\behat\Context
```
I am ok with this change as `ContextDump` is not much helpful with AJAX...

ibelar replied:
ContextDump is useful when testing locally though.

mvorisek replied:
Yes, but static page issues are discovered even with phpunit... Marking as resolved.

mvorisek commented on behat.yml.dist:
```diff
@@ -20,4 +20,4 @@ default:
                   args:
                     - '--headless'
-                    - '--window-size=1280,720'
+                    - '--window-size=1920,1080'
```
I think we should solve this in Behat testing (I think focus is needed), otherwise tests may fail when testing locally with smalled viewport.

ibelar replied:
From what I could see, --window-size switch is used when --headless mode is on only.
Then if window size is set too small in --headless mode, then we got 'button not find 'or 'content not find' errors.
Does not happen when --headless is off.

mvorisek replied:
In non-headless mode, we can not expected that the window will be full HD as well.

We want to test in smaller screen in general (to test if even focusable/scrollable in smaller devices).

ibelar replied:
In non-headless, the test pass, no matter what size you set --window-size.
In headless mode, size matters because it seems like window is not scrolling using that mode, which make the test failed. 

If you want to test focusable/scrollable, we should not use headless mode at all.

mvorisek replied:
I just tried to resize non-headless Chrome local browser to size like 200x200 and 3 tests failed:
![image](https://user-images.githubusercontent.com/2228672/98271974-d426db00-1f90-11eb-9f33-6eb09219eaff.png)

I will test this more...

ibelar replied:
If I am not using --headless (comment it out) and used window-size of 200,200 then callback feature test pass for me locally. It fail using --headless option.

Pass:
                switches:
                  - "--window-size=200,200"

Fail:
                switches:
                  - "--headless"
                  - "--window-size=200,200"

mvorisek replied:
https://github.com/atk4/ui/compare/behat_size_debug / https://github.com/atk4/ui/runs/1359529478 with reverted size, it seems to fail for another reason (empty expected text, not bebause something is not clickable)

ibelar replied:
Not using window size option make test pass for headless or non-headless.
May be we should remove "--window-size" completely.

ibelar replied:
> https://github.com/atk4/ui/compare/behat_size_debug / https://github.com/atk4/ui/runs/1359529478 with reverted size, it seems to fail for another reason (empty expected text, not bebause something is not clickable)

yes - exactly why I have increase size. Because the container using 'word' is below the window size if reverted.

mvorisek replied:
because if may use larger size... Current settings are good I think, as it defines size for screenshotting etc.

ibelar replied:
Current settings is clearly an issue with headless mode. Besides there is no need for screenshots using headless mode. 
I suggest we remove, or increase, window-size value. 
Local setting can be overrides as well.

mvorisek replied:
I am strictly againt it. This does not address real issue it only increases size.

The sandpit test above also shown, that very unrelated message is shown.

ibelar replied:
there is definitely an issue in Selenium --headless mode with window size. My guess, only a guess, is when we reload content via callback then something is going with browser window an element text is not accessible after callback.
I have issue a workaround for it. See https://github.com/atk4/ui/pull/1469/commits/59f73d38dc0e905b41aa2e3c1102ecf6cb04a8dd

mvorisek replied:
I also wonder why there in no Behat "Stack trace:"...

ibelar replied:
I have include #1538  configuration in this pr for the test to pass temporarily. Lets' resolve that issue over there.

mvorisek commented on tests-behat/bootstrap/Context.php:
```diff
@@ -419,6 +430,139 @@ public function iClickIconUsingCss($arg1)
+     */
+    public function iCheckIfInputMatch($name)
+    {
+        $expected = $this->getSession()->getPage()->find('css', '.atk-expected-input-result .content')->getText();
```
I do not see any advantage of this, we should just `I check if input XXX match YYY`

ibelar replied:
Because it will test if ScopeBuilder is properly building Vue Query Builder (VQB) query. VQB json query object is set by ScopeBuilder using Model condition. It should match value predefine by condition set in test file.

mvorisek replied:
If you pass the target as selector (like in the next comment), I am ok with it, but Context class should not grow without significant reason for new API.

mvorisek commented on tests-behat/bootstrap/Context.php:
```diff
@@ -419,6 +430,139 @@ public function iClickIconUsingCss($arg1)
+    }
+
+    /**
+     * @Then /^I check if word from data scope match$/
```
The same here, just add the expected text to feature.

mvorisek replied:
or add at least that two selectors as args, ie. check if elem(arg1) value === elem(arg2) value

ibelar replied:
I added it to the demo file because we can change it without changing the Context file.
Will try to set up using the argument.

ibelar committed:
fixes

mvorisek commented on tests-behat/bootstrap/Context.php:
```diff
@@ -482,41 +482,41 @@ public function scopeBuilderBoolRule($name, $value)
 
-        if (preg_replace('/\s*/m', '', $expectedWord) !== preg_replace('/\s*/m', '', $responseWord)) {
-            throw new \Exception('Data word does not match: ' . $responseWord . ' expected: ' . $expectedWord);
+        if (preg_replace('~\s*~', '', $expectedText) !== preg_replace('~\s*~', '', $compareToText)) {
```
When expected test is generated, when it does not match binary? I think, we do not need this replace... 2x

ibelar replied:
done

mvorisek replied:
also for `compareInputValueToElementText`

ibelar replied:
test fail without preg_replace for input...

mvorisek replied:
then we expect wrong input ;-) - must be fixed

ibelar replied:
Both, $expected and $input-getValue(), contains space char. $expected for readability and $get->input() for "matches regular expression" so we need to remove space in both place for test to pass.

ibelar committed:
remve pregreplace for word comparaison

mvorisek committed:
no ui exception in Behat

ibelar committed:
workaround for window-size

ibelar committed:
Merge branch 'refactor/calendar' of https://github.com/atk4/ui into refactor/calendar

mvorisek commented on behat.yml.dist:
```diff
@@ -20,4 +20,4 @@ default:
                   args:
                     - '--headless'
-                    - '--window-size=1280,720'
+                    - "--window-size=1280,720"
```
```suggestion
                    - '--window-size=1280,720'
```

mvorisek replied:
no change, as before...

ibelar committed:
Update behat.yml.dist

Co-authored-by: Michael Voříšek <mvorisek@mvorisek.cz>

ibelar commented:
> no, definitely no, we need to address the real issue

Do you refer to the workaround? If so, then I will simply remove the test for word, because this is only the way I could find beside removing --window-size option...

ibelar commented:
BTW - why is Behat(latest, Chrome Lowest) take so much time?

mvorisek commented:
> 
> 
> BTW - why is Behat(latest, Chrome Lowest) take so much time?

because it resolves dependencies, maybe because of composer2, maybe because some dep "expends to large dep tree to explore", I have seen it, I might help later or/and check with composer guys

mvorisek commented:
> 
> 
> > no, definitely no, we need to address the real issue
> 
> Do you refer to the workaround? If so, then I will simply remove the test for word, because this is only the way I could find beside removing --window-size option...

The fix is horrible in sense of cleaness, so that fix defitely no, I still do not know the real issue, exception is unrelated - the real issue is silence probably as there is JS in between which does not throw...

mvorisek commented:
> 
> 
> BTW - why is Behat(latest, Chrome Lowest) take so much time?

let's wait until https://github.com/composer/composer/pull/9261 is released, it should solve exactly this issue

ibelar committed:
Merge branch 'develop' into refactor/calendar

ibelar committed:
update js release

ibelar committed:
Merge branch 'develop' into refactor/calendar

ibelar committed:
new arrow function usage

ibelar committed:
Merge branch 'develop' into refactor/calendar

ibelar committed:
remove workaround

ibelar committed:
fix behat

- fix until #1538 is resolved

ibelar committed:
add temporary fix

until #1538 is resolve

ibelar committed:
cs fix

ibelar committed:
revert headless with bigger window size

DarkSide666 reviewed (approved):
LGTM. Thanks for a hard work!
